psi4/driver/procrouting/diis.py

```diff
@@ -181,33 +232,170 @@ def extrapolate(self, *args):
+
+    def ediis_populate(self):
+        num_entries = len(self.stored_vectors)
+
+        self.ediis_quadratic = np.zeros((num_entries, num_entries))
+        for i in range(num_entries):
+            for item_num in range(len(self.template["densities"])):
+                d = self.load_quantity("densities", i, item_num)
+                for j in range(num_entries):
+                    f = self.load_quantity("target", j, item_num)
+                    self.ediis_quadratic[i][j] += d.vector_dot(f)
+
+        diag = np.diag(self.ediis_quadratic)
+        # D_i F_i + D_j F_j - D_i F_j - D_j F_i; First two terms use broadcasting tricks
+        self.ediis_quadratic = diag[:, None] + diag - self.ediis_quadratic - self.ediis_quadratic.T
+
+        self.ediis_quadratic *= -1/2
+
+        if self.closed_shell:
+            self.ediis_quadratic *= 2
+
+    def extrapolate(self, *args, Dnorm = None):
+        """ Perform extrapolation. Must be passed in the RMS error to decide how to handle hybrid algorithms. """
+
+        if {"adiis", "ediis"}.intersection(self.engines) and Dnorm is None:
+            raise ValidationError("An extrapolation engine insists you specify the error metric.")
+
+        performed = set()
+
+        if self.engines == {"diis"}:
+            coeffs = self.diis_coefficients()
+            performed.add("DIIS")
+        elif len(self.engines) == 1:
+            blend_start = core.get_option("SCF", "INITIAL_SCF_BLEND_START")
+            blend_stop = core.get_option("SCF", "INITIAL_SCF_BLEND_STOP")
+            if Dnorm <= blend_stop:
+                return performed
+            elif self.engines == {"ediis"}:
+                coeffs = self.ediis_coefficients()
+                performed.add("EDIIS")
+            elif self.engines == {"adiis"}:
+                coeffs = self.adiis_coefficients()
+                performed.add("ADIIS")
+            else:
+                raise Exception(f"DIIS engine not recognized: {self.engines[0]}.")
+        elif self.engines == {"diis", "adiis"} or self.engines == {"diis", "ediis"}:
+            blend_start = core.get_option("SCF", "INITIAL_SCF_BLEND_START")
+            blend_stop = core.get_option("SCF", "INITIAL_SCF_BLEND_STOP")
+            if "adiis" in self.engines:
+                initial_coefficient_function = self.adiis_coefficients
+                initial_name = "ADIIS"
+            else:
+                initial_coefficient_function = self.ediis_coefficients
+                initial_name = "EDIIS"
+
+            if Dnorm >= blend_start:
+                coeffs = initial_coefficient_function()
+                performed.add(initial_name)
+            elif Dnorm <= blend_stop:
+                coeffs = self.diis_coefficients()
+                performed.add("DIIS")
+            else:
+                m = 1 - (Dnorm - blend_start) / (blend_stop - blend_start)
+                coeffs = m * initial_coefficient_function() + (1 - m) * self.diis_coefficients()
```

Comment:
Not sure about the keyword naming: `INITIAL_SCF_BLEND_START` and `INITIAL_SCF_BLEND_STOP` control when to start mixing in DIIS and when to switch over fully to DIIS. IIRC the default values are 1e-2 and 1e-3, respectively.

Reply:
I share your concern, so let me know if you think of better names. I've gone through multiple iterations on the names of these keywords and I don't expect I'll come up with anything better.

Reply:
Also, 1e-1 and 1e-4 according to the Garza and Scuseria paper.

Reply:
What about `INITIAL_DIIS_THRESHOLD` to start mixing in DIIS and `DIIS_THRESHOLD` to fully switch over?

Reply:
> Also, 1e-1 and 1e-4 according to the Garza and Scuseria paper.

OK. Might be worthwhile to put in the reference for the default values

Reply:
The meaning of `DIIS_THRESHOLD` is unclear from the name. It isn't even clear that it has anything to do with ADIIS/EDIIS.

Reply:
> > Also, 1e-1 and 1e-4 according to the Garza and Scuseria paper.
> 
> OK. Might be worthwhile to put in the reference for the default values

Fine by me. I can add this as a comment in the options file, but I don't think this should be user-facing, as it's too technical.

Reply:
I would suggest `INITIAL_SCF_ACCELERATOR_START/STOP` to keep the names similar to the parent option.

Reply:
Those names are misleading: it's not the A/EDIIS that starts at the first keyword, but the blending in of `DIIS`. `INITIAL_SCF_ACCELERATOR_BLEND_START/STOP` is more descriptive, but also more verbose. I'll wait for more opinions before deciding if the tradeoff is worth it.